### PR TITLE
Java console app (Device code flow sample)

### PR DESCRIPTION
**Pull request by ramya25**

### CODE REVIEW

File: src/main/java/DeviceCodeFlow.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Consumer;
+
+public class DeviceCodeFlow {
+
+    final static String PUBLIC_CLIENT_ID = "de165c80-4281-47c5-b23b-42ec6080db52";
```
Comment (navyasric) [inline src/main/java/DeviceCodeFlow.java]: Do we want to have a client id hard coded here? We can just use placeholder text.

File: README.md [hunk 2]
```diff
@@ -1,64 +1,199 @@
...
+- And call to a web API (in this case, the [Microsoft Graph](https://graph.microsoft.com))
+
+To emulate a device not capable of showing UX, the sample is packaged as a Java console application.
+The application signs-in users with Azure Active Directory (Azure AD), using the Microsoft Authentication Library for Java (MSAL4J) to obtain a JWT access token through the OAuth 2.0 protocol. The access token is then used to call the Microsoft Graph API to obtain information about the user who signed-in. The sample is structured so that you can call your own API
+
+![Topology](./ReadmeFiles/Topology.png)
+
+### Scenario
+
+The application obtains tokens through a two steps process especially designed for devices and operating systems that cannot display any UX. Examples of such applications are applications running on iOT, or Command-Line tools (CLI). The idea is that:
```
Comment (navyasric) [inline README.md]: typo: IOT

File: README.md [hunk 3]
```diff
@@ -1,64 +1,199 @@
...
+
+1. whenever a user authentication is required, the command-line app provides a code and asks the user to use another device (such as an internet-connected smartphone) to navigate to [https://microsoft.com/devicelogin](https://microsoft.com/devicelogin), where the user will be prompted to enter the code. That done, the web page will lead the user through a normal authentication experience, including consent prompts and multi factor authentication if necessary.
```
Comment (navyasric) [inline README.md]: typo: uppercase W in Whenever
reword: Once that is done, the web page ....

File: README.md [hunk 4]
```diff
@@ -1,64 +1,199 @@
...
+   - Click the **Add a permission** button and then,
+   - Ensure that the **Microsoft APIs** tab is selected.
+   - In the *Commonly used Microsoft APIs* section, click on **Microsoft Graph**
+   - In the **Delegated permissions** section, select the **User.Read**, **User.ReadBasic.All** in the list. Use the search box if necessary.
+   - Click on the **Add permissions** button at the bottom.
+
+1. At this stage, the permissions are assigned correctly but since the client app does not allow users to interact, the user's themselves cannot consent to these permissions.
+   To get around this problem, we'd let the [tenant administrator consent on behalf of all users in the tenant](https://docs.microsoft.com/azure/active-directory/develop/v2-admin-consent).
+   Click the **Grant admin consent for {tenant}** button, and then select **Yes** when you are asked if you want to grant consent for the requested permissions for all account in the tenant.
+   You need to be the tenant admin to be able to carry out this operation.
```
Comment (navyasric) [inline README.md]: Is this step really necessary? If the tenant admin doesn't grant consent, won't the user get a chance to consent when performing the device code auth flow on the phone browser? (that's what we say in the scenario overview)

File: README.md [hunk 5]
```diff
@@ -1,64 +1,199 @@
...
+page_type: sample
+languages:
+  - java  
+products:
+  - azure
+  - azure-active-directory  
+  - office-ms-graph
+description: "This sample demonstrates a java console application calling The Microsoft Graph"
+---
+# Invoking an API protected by Azure AD from a text-only device
```
Comment (navyasric) [inline README.md]: Invoking an API protected by Azure AD from a device without a browser for interaction

File: README.md [hunk 6]
```diff
@@ -1,64 +1,199 @@
...
+service: Microsoft Graph
+endpoint: Microsoft identity platform
+page_type: sample
+languages:
+  - java  
+products:
+  - azure
+  - azure-active-directory  
+  - office-ms-graph
+description: "This sample demonstrates a java console application calling The Microsoft Graph"
```
Comment (navyasric) [inline README.md]: This sample demonstrates calling the Microsoft Graph from a Java console application without a browser for interaction

**Review by navyasric — APPROVED** [review]
Looks good. Added a few suggestions

**Review by kalyankrishna1 — APPROVED** [review]

File: src/main/java/DeviceCodeFlow.java (hunk 1 above)
Comment (ramya25) [reply src/main/java/DeviceCodeFlow.java]: Yes. Committed with client id by mistake, will remove it.

File: README.md (hunk 3 above)
Comment (ramya25) [reply README.md]: Updated

File: README.md [hunk 7]
```diff
@@ -1,64 +1,199 @@
...
+- Acquire an [Access token](https://docs.microsoft.com/azure/active-directory/develop/access-tokens) for a web API (in this case, the [Microsoft Graph](https://graph.microsoft.com))
+
+To emulate a an app on a device that is not capable of showing a UX, the sample is packaged as a Java console application.
+The application signs-in users with Azure Active Directory (Azure AD), using the Microsoft Authentication Library for Java (MSAL4J) to obtain a JWT access token through the OAuth 2.0 protocol. The access token is then used to call the Microsoft Graph API to obtain information about the user who signed-in. The sample is structured as such so that you can call your own API.
+
+![Topology](./ReadmeFiles/Topology.png)
+
+### Scenario
+
+The application obtains tokens through a two step process especially designed for devices and operating systems that cannot display any UX. Examples of such applications are applications running on iOT, or Command-Line tools (CLI). The idea is that:
```
Comment (sangonzal) [inline README.md]: nit: I think the way this is commonly capitalized is IoT

File: README.md [hunk 8]
```diff
@@ -1,64 +1,199 @@
...
+
+1. Whenever a user is required to authenticate, the command-line app provides a code and asks the user to use another device (such as an internet-connected smartphone) to navigate to [https://microsoft.com/devicelogin](https://microsoft.com/devicelogin), where the user will be prompted to enter the code provided earlier. Oce the user enters the code, the web page will lead the user through a normal authentication experience, including presenting any consent prompts and take the user through a multi-factor authentication experience if necessary.
```
Comment (sangonzal) [inline README.md]: Once*

"...and taking* the user through a multi-factor..."

File: README.md [hunk 9]
```diff
@@ -1,64 +1,199 @@
...
+
+1. Upon successful authentication, the command-line app will receive the required tokens through a back channel and will use it to perform the web API calls it needs. In this case, the sample retrieves a list os users in the signed-in users tenant from Microsoft Graph and lists that in the window.
```
Comment (sangonzal) [inline README.md]: "of* users in the signed-in"

"and lists it in the window"*

**Review by sangonzal — APPROVED** [review]
Thanks, looks good to me

File: README.md (hunk 9 above)
Comment (ramya25) [reply README.md]: updated

File: README.md (hunk 8 above)
Comment (ramya25) [reply README.md]: corrected

File: README.md (hunk 7 above)
Comment (ramya25) [reply README.md]: Yes updated.

File: README.md (hunk 6 above)
Comment (ramya25) [reply README.md]: updated

File: README.md (hunk 5 above)
Comment (ramya25) [reply README.md]: updated

File: README.md (hunk 2 above)
Comment (ramya25) [reply README.md]: Updated as IoT

File: README.md (hunk 4 above)
Comment (ramya25) [reply README.md]: Will Check and update

Comment (ramya25) [reply README.md]: This step is necessary

Comment (ramya25) [reply README.md]: Tested several times and now removed this step. I think `error: Forbidden` comes when I am testing for first time in mac, but when I try to login first time in windows machine and then in mac, the error doesn't show up any where.

Comment (ramya25) [reply README.md]: changed the scope in the code. so this issue solved.

**Pull request merged**